jwst/ramp_fitting/ramp_fit.py

```diff
@@ -549,79 +537,332 @@ def ols_ramp_fit(data, err, groupdq, inpixeldq, buffsize, save_opt, readnoise_2d
 
-        # If all groups were flagged, the final group would have been picked up
-        #   in the while loop above, ngroups would have been set to 0, and Nones
-        #   would have been returned.  If execution has gotten here, there must
-        #   be at least 1 remaining group that is not all flagged.
-        last_gdq = groupdq[:, -1, :, :]
-        if np.all(np.bitwise_and(last_gdq, DO_NOT_USE)):
-            nreads -= 1
-            ngroups -= 1
+def ramp_fit_miri_multigroups(data, err, groupdq, cubeshape, imshape, nreads, ngroups):
+    """
+    For MIRI datasets having >1 group, if all pixels in the final group are
+    flagged as DO_NOT_USE, resize the input model arrays to exclude the
+    final group.  Similarly, if leading groups 1 though N have all pixels
+    flagged as DO_NOT_USE, those groups will be ignored by ramp fitting, and
+    the input model arrays will be resized appropriately. If all pixels in
+    all groups are flagged, return None for the models.
 
-            # Check if there are remaining groups before accessing data
-            if ngroups < 1:  # no usable data
-                log.error('2. All groups have all pixels flagged as DO_NOT_USE,')
-                log.error('  so will not process this dataset.')
-                return None, None, None
+    Parameters
+    ----------
+    data: ndarray
+        4-D image cube with dimensions (integrations, groups, rows, columns)
 
-            data = data[:, :-1, :, :]
-            err = err[:, :-1, :, :]
-            groupdq = groupdq[:, :-1, :, :]
-            cubeshape = (nreads,) + imshape
+    err: ndarray
+        Error array
 
-            log.info('MIRI dataset has all pixels in the final group flagged as DO_NOT_USE.')
+    groupdq: ndarray
+        GroupDQ array
 
-        # Next block is to satisfy github issue 1681:
-        # "MIRI FirstFrame and LastFrame minimum number of groups"
-        if ngroups < 2:
-            log.warning('MIRI datasets require at least 2 groups/integration')
-            log.warning('(NGROUPS), so will not process this dataset.')
-            return None, None, None
+    cubeshape: tuple
+        Image cube dimensions per integration
 
-    if ngroups == 1:
-        log.warning('Dataset has NGROUPS=1, so count rates for each integration')
-        log.warning('will be calculated as the value of that 1 group divided by')
-        log.warning('the group exposure time.')
+    imshape: tuple
+        Image shape per group
 
-    # Calculate effective integration time (once EFFINTIM has been populated
-    #   and accessible, will use that instead), and other keywords that will
-    #   needed if the pedestal calculation is requested. Note 'nframes'
-    #   is the number of given by the NFRAMES keyword, and is the number of
-    #   frames averaged on-board for a group, i.e., it does not include the
-    #   groupgap.
+    nreads: int
```

Comment:
Can this be derived from `data.shape` as well?

Reply:
Yes. That's part of what I'm currently refactoring. The primary focus of this PR is to break down ols_ramp_fit into a more manageable size calling sub-functions. The follow on PR's that will come will contain the type of refactoring you are describing, simplifying the variables used and passed around.